dags/mag_and_friends.py

```diff
@@ -426,22 +435,30 @@ def update_provenance():
                                     # Write back to the file
                                     writer.writerows(all_data.values())
 
+                            update_provenance = PythonOperator(
+                                task_id="update_provenance",
+                                python_callable=update_provenance_callable,
+                                task_group=current_tasks
+                            )
+
                             copy_study_to_cluster >> \
                             run_mag >> \
                             watch_mag >> \
-                            copy_results_from_cluster >> \
-                            post_process_results() >> \
-                            update_provenance()
+                            update_provenance
 
                             copy_study_to_cluster >> \
                             run_metatdenovo >> \
                             watch_metatdenovo >> \
-                            copy_results_from_cluster 
+                            copy_metatdenovo_results_from_cluster >> \
+                            post_process_results >> \
+                            update_provenance
                             
                             copy_study_to_cluster >> \
                             run_taxprofiler >> \
                             watch_taxprofiler >> \
-                            copy_results_from_cluster
+                            copy_taxprofiler_results_from_cluster >> \
+                            post_process_results >> \
+                            update_provenance
 
                             if os.path.exists(accessionsFile):
                                 copy_study_to_cluster >> \
```

Comment:
In addition, do you think there will be good to have some trigger rules in place to ensure tasks progress as desired, and skip all upon failure at specific steps?

Reply:
not sure what you mean here tbh. what trigger rules did you have in mind?

Reply:
For dynamic task mapping I had to enforce TriggerRule.ALL_SUCCESS to make sure tasks are executed in a sequential order.  Guess it is not needed for custom looping setup

Reply:
that shouldnt be necessary. i looked at your PR, and commented there on why i think you had to do this. if unclear, let me know.